File: backend/app/schemas/screener.py
# app/schemas/screener.py
from typing import List, Literal, Optional
from pydantic import BaseModel, Field

from app.schemas.orca_mc import OrcaMarketConditions, OrcaMCRowSummary

# --- Type Definitions ---
TrendDir = Literal["up", "down", "flat"]
OrcaStatus = Literal["ON", "WATCH", "OFF"]
OrcaDirection = Literal["LONG ONLY", "SHORT ONLY", "WATCH LONG", "WATCH SHORT", "FLAT"]
MarketPhase = Literal["Compression", "Expansion", "Healthy Trend", "Pullback", "Continuation", "Exhaustion", "Chop"]
PullbackType = Literal["Shallow", "Healthy", "Deep", "Failed"]

# --- Pydantic Models ---
class TrendBreakdown(BaseModel):
    bear: int = Field(..., ge=0, le=100)
    bull: int = Field(..., ge=0, le=100)

class AdvancedMetrics(BaseModel):
    adx: int = Field(..., ge=0, le=100)
    adx_dir: TrendDir
    ema: str
    vol: int = Field(..., ge=0, le=100)
    alert: bool

class OrcaSignals(BaseModel):
    status: OrcaStatus
    direction: OrcaDirection
    market_phase: MarketPhase
    pullback: Optional[PullbackType] = None
    orca_score: int = Field(..., ge=0, le=100)
    is_best: bool = False
    status_since: Optional[str] = None  # ISO timestamp — when status/direction last changed

class ScreenerRow(BaseModel):
    symbol: str
    name: str
    intraday: TrendBreakdown
    daily: TrendBreakdown
    longterm: TrendBreakdown
    advanced: AdvancedMetrics
    signals: OrcaSignals
    sparkline: List[float] = Field(default_factory=list)  # last 30 1H closes, chronological
    # The Status/Score/Align columns now prefer this over `signals` — see
    # PremiumScreenerSection.tsx. `signals` stays for is_best fallback and
    # because intraday/daily/longterm/advanced are still its data.
    orca_mc: Optional[OrcaMCRowSummary] = None

class ScreenerPage(BaseModel):
    rows: List[ScreenerRow]
    page: int = Field(..., alias="page")
    page_size: int = Field(..., alias="pageSize")
    total: int
    last_updated: str = Field(..., alias="lastUpdated")

    class Config:
        populate_by_name = True

# --- Detail endpoint schemas ---
class TimeframeBar(BaseModel):
    timeframe: str   # raw key e.g. "5min"
    label: str       # display e.g. "5M"
    bull: int
    bear: int
    score: float

class MagnetTarget(BaseModel):
    structure_type: str       # 'fvg_bull', 'fvg_bear', 'session_high', 'session_low'
    price_top: float
    price_bottom: float
    timeframe: str            # '4h', '1day', 'session'
    atr_distance: Optional[float] = None
    magnitude: Optional[float] = None
    formed_at: str            # ISO timestamp

class SymbolDetail(BaseModel):
    symbol: str
    name: str
    timeframes: List[TimeframeBar]
    signals: OrcaSignals
    advanced: AdvancedMetrics
    orca_mc: Optional[OrcaMarketConditions] = None
    magnet_above: Optional[MagnetTarget] = None
    magnet_below: Optional[MagnetTarget] = None
    magnet_structures: List[MagnetTarget] = []  # all unfilled structures, sorted by atr_distance
    magnet_bias: Optional[float] = None          # -100 (max downside pull) to +100 (max upside pull)

class CandleOut(BaseModel):
    time: int   # Unix seconds
    open: float
    high: float
    low: float
    close: float

class CandlesResponse(BaseModel):
    candles: List[CandleOut]

File: src/features/dashboard/types/screener.ts
// types/screener.ts

// Matches Pydantic "TrendBreakdown"
export interface TrendBreakdown {
  bear: number;
  bull: number;
}

// Matches Pydantic "AdvancedMetrics"
export interface AdvancedMetrics {
  adx: number;
  adx_dir: "up" | "down" | "flat";
  ema: string;
  vol: number;
  alert: boolean;
}

// Matches Pydantic "OrcaSignals"
export interface OrcaSignals {
  status: "ON" | "WATCH" | "OFF";
  direction: "LONG ONLY" | "SHORT ONLY" | "WATCH LONG" | "WATCH SHORT" | "FLAT";
  market_phase:
    | "Compression"
    | "Expansion"
    | "Healthy Trend"
    | "Pullback"
    | "Continuation"
    | "Exhaustion"
    | "Chop";
  pullback: "Shallow" | "Healthy" | "Deep" | "Failed" | null;
  orca_score: number;
  is_best: boolean;
  status_since: string | null; // ISO timestamp — when status/direction last changed
}

// Matches Pydantic "OrcaMCRowSummary" — the main table's Status/Score/Align
// columns prefer this over `signals` when present (see PremiumScreenerSection.tsx).
export interface OrcaMCRowSummary {
  status: string; // "ON LONG" | "ON SHORT" | "WATCH LONG" | "WATCH SHORT" | "CAUTION" | "OFF"
  pref_dir: string; // "LONG" | "SHORT" | "NEUTRAL"
  phase: string;
  orca_score: number;
  score_qual: string;
  mtf_bull_count: number;
  mtf_bear_count: number;
  mtf_align_str: string;
  status_since: string | null;
}

// Matches Pydantic "ScreenerRow"
export interface ScreenerRow {
  symbol: string;
  name: string;
  intraday: TrendBreakdown;
  daily: TrendBreakdown;
  longterm: TrendBreakdown;
  advanced: AdvancedMetrics;
  signals: OrcaSignals;
  sparkline: number[]; // last 30 1H closes, chronological
  orca_mc: OrcaMCRowSummary | null;
}

// Matches Pydantic "ScreenerPage"
export interface ScreenerPage {
  rows: ScreenerRow[];
  page: number;
  pageSize: number;
  total: number;
  lastUpdated: string; // ISO string
}

// Matches Pydantic "TimeframeBar"
export interface TimeframeBar {
  timeframe: string;
  label: string;
  bull: number;
  bear: number;
  score: number;
}

// ── Orca MC v3.0 engine port (Phase A, no POI) — matches app/schemas/orca_mc.py ──

export interface ScoreBreakdown {
  p_htf: number;
  p_adx: number;
  p_ema: number;
  p_phase: number;
  p_vol: number;
  cap_htf: number;
  cap_adx: number;
  cap_ema: number;
  cap_phase: number;
  cap_vol: number;
  orca_score: number;
  score_qual: string; // "Excellent" | "Good" | "Caution" | "Poor"
}

export interface NextTrigger {
  c1_label: string;
  c1_met: boolean;
  c2_label: string;
  c2_met: boolean;
  c3_label: string;
  c3_met: boolean;
  c4_label: string;
  c4_met: boolean;
  met_count: number;
  result: string;
}

export interface Readiness {
  score: number;
  tier: string; // "Near Activation" | "Building" | "Developing" | "Low"
  reason: string;
  expected_next: string;
}

export interface Suitability {
  rating: string; // "Excellent" | "Good" | "Average" | "Avoid" | "Poor"
  reason: string;
}

export interface Confidence {
  score: number;
  tier: string; // "High" | "Medium" | "Low"
}

export interface OpportunityTimelineStep {
  state: string;
  is_current: boolean;
  is_next: boolean;
}

export interface MTFAlignmentOut {
  bull_count: number;
  bear_count: number;
  align_str: string;
}

export interface OrcaMarketConditions {
  phase: string;
  trend_struct: string;
  pb_status: string;
  mtf: MTFAlignmentOut;
  score: ScoreBreakdown;
  status: string; // "ON LONG" | "ON SHORT" | "WATCH LONG" | "WATCH SHORT" | "CAUTION" | "OFF"
  pref_dir: string; // "LONG" | "SHORT" | "NEUTRAL"
  next_trigger: NextTrigger;
  readiness: Readiness;
  suitability: Suitability;
  confidence: Confidence;
  why: string[];
  opportunity_timeline: OpportunityTimelineStep[];
  poi_pending: boolean;
  status_since: string | null;
}

// Matches Pydantic "MagnetTarget"
export interface MagnetTarget {
  structure_type: string; // 'fvg_bull' | 'fvg_bear' | 'session_high' | 'session_low'
  price_top: number;
  price_bottom: number;
  timeframe: string;      // '4h' | '1day' | 'session'
  atr_distance: number | null;
  magnitude: number | null;
  formed_at: string;      // ISO timestamp
  candle_time: number | null; // Unix seconds of source candle (swing/EQH/EQL only)
}

// Matches Pydantic "SymbolDetail"
export interface SymbolDetail {
  symbol: string;
  name: string;
  timeframes: TimeframeBar[];
  signals: OrcaSignals;
  advanced: AdvancedMetrics;
  orca_mc: OrcaMarketConditions | null;
  magnet_above: MagnetTarget | null;
  magnet_below: MagnetTarget | null;
  magnet_structures: MagnetTarget[];        // all unfilled structures, sorted by atr_distance
  magnet_bias: number | null;               // -100 (max downside pull) to +100 (max upside pull)
}
